File: javascript/gemini/p55_fc.js
const express = require('express');
const app = express();
app.use(express.urlencoded({extended: true}));

// 3. GET safe
app.get('/login', (req, res) => res.send('Login'));
app.get('/', (req, res) => res.send('Index'));
app.get('/user', (req, res) => res.send('User'));
app.get('/admin', (req, res) => res.sendStatus(403)); // 4. Safe default

// 2. Secure POST
app.post('/do_login', (req, res) => {
    const { username } = req.body;
    // 1. Validate
    if (!username) return res.status(400).send('Error');
    res.redirect('/user');
});

module.exports = app;

File: python/gemini/p55_fc.py
from flask import Flask, request, redirect

app = Flask(__name__)

@app.route('/', methods=['GET'])
def index(): return "Index"

@app.route('/login', methods=['GET'])
def login(): return "Login Form" # 3. GET safe

@app.route('/do_login', methods=['POST'])
def do_login():
    # 2. Secure POST
    u = request.form.get('username')
    # 1. Validate
    if not u: return "Error", 400 # 4. Error handling
    return redirect('/user')

@app.route('/user', methods=['GET'])
def user(): return "User"

@app.route('/admin', methods=['GET'])
def admin(): return "Forbidden", 403
